### ISSUE TITLE
Backup/Restore (onto new devices)

### ISSUE THREAD

**Issue by GoProAndy** [mattrubin/authenticator#6]
Hi,
sorry for using this tool, but what is happens if I get a new iPhone? Is the content also lost after Restoring the iCloud Backup? Till now i use the the Google Authenticator app and here is all lost. 
Thanks
Andreas

**Comment by PatTheMav** [comment]
After Apple "had" to give me a new iPhone (after they switched my display in the Apple Store but couldn't calibrate it), I lost all my 2FA tokens (because I could only to an iCloud Backup before the repair). So storing the necessary information in the iCloud Keychain would be great (apparently "LockDown" uses it, Authy uses their own infrastructure unfortunately).

So this is definitely a +1 from me, especially now that @mattrubin seems to be porting the whole thing to Swift (massive amounts of commits happened in the past few days).

I still prefer this app over others because of it's simple, straight-forward design and high speed (and being open source), so iCloud Keychain sync would definitely be a great addition.

**Comment by jakeswenson** [comment]
I also love this app, and would love to see the OTP secrets stored in the iCloud keychain.
@mattrubin Thoughts?

**Comment by SurajGupta** [comment]
This would be THE feature that would kill Google Authenticator for good.  Switching devices or reinstalling the OS and restoring data on the same device are painful experiences right now because of the need to re-establish TFA accounts.

Users have been asking Google for this since 2013:
https://groups.google.com/forum/#!topic/google-authenticator-issues/baROYFInQk4

**Comment by SurajGupta** [comment]
@mattrubin - thoughts on my last comment?

**Comment by putorti** [comment]
According to Apple, all iCloud backups are encrypted. However my last two restores have lost all my tokens. I may be mistaken but I do believe it worked at one point, but I've definitely lost all my tokens twice. My backup is 100% complete. Seems like a possible bug to me.

**Comment by putorti** [comment]
Also, maybe something is wrong on my end but I see the last update of this app on the App Store as September 29, 2014. So no updates in nearly two years.

**Comment by jakeswenson** [comment]
@putorti There has been a significantly awesome rewrite in progress to swift from objc. So i'd imagine thats why the long period of no updates. I'm not sure where @mattrubin has gone though :)
it'd be awesome to get an update, maybe i'm just looking in the wrong spot.

**Comment by beaucollins** [comment]
My vote is an encrypted export text file that initiates the share sheet (and the requisite import).

This allows me to get to my secrets while also letting me control exactly where they will end up.

**Comment by joshperry** [comment]
It looks like setting the [kSecAttrSynchronizable](https://developer.apple.com/reference/security/ksecattrsynchronizable) key will automatically add a keychain item to the sync list.

However, the first caveat makes me wonder if this is wise as these OTP schemes are meant as proxies for hardware tokens and seem like they should be unique per device so that they can be managed (revoked) at device granularity. Does it make sense to have the same token on multiple devices?

>Updating or deleting items using the kSecAttrSynchronizable key will affect all copies of the item, not just the one on your local device. Be sure that it makes sense to use the same password on all devices before making a password synchronizable.

**Comment by PatTheMav** [comment]
So here's what I saw another app doing: It allowed me to export my tokens as QR codes via the print function. That way I was able to "port" them into Authenticator by re-scanning those QR codes.

Secured behind Touch/Face ID this might be worth looking into.

Having used the iOS 11 betas, this saved me a lot (!) of headaches..

**Comment by ghost** [comment]
What about the people who need to, or even would like to, "reset" their iOS device for some reason? This app makes it impossible without losing all 2FA information. I have not seen another app with this restriction.

How about adding an option to, securely via iTunes, export an archive of the tokens? (iOS 11 requires a passcode to trust a computer now, so this option would prove more secure than in the past.) :)

Edit: That backup could then be securely imported via AirDrop/iTunes etc. to the new iOS install.

@mattrubin thoughts?

**Comment by jakeswenson** [comment]
I agree with @PatTheMav. Even just give me the ability to see the token value, some how. 

I understand the "uniqueness" argument, but give me some way of making my own decisions about how a value my HOTP tokens. (A button that says are you really sure you want to see the token?) 

As of right now I get around this issue by not using QR codes anymore, I just tell the website to give me the token string and then I store it in my password manager and manually setup my authenticator tokens.

**Comment by metbril** [comment]
Since iOS 11 Apple allows backup of Health data in iCloud backups. 

https://www.imore.com/how-sync-your-health-data-ios-11-and-how-it-works

Would now be the time to backup the tokens, too?

**Comment by cosmoburn** [comment]
Having recently lost my phone and having gone through a nightmare to get all my keys back what @metbril suggested is the same thing my friend who is an iOS dev said. This feature would keep me using this app.

**Comment by csaftoiu** [comment]
Why can't you display a QR code on old phone, and scan it from new phone to transfer everything?

**Comment by csaftoiu** [comment]
itunes doesn't exist anymore :((( how can i transfer to new phone?

**Comment by beaucollins** [comment]
> itunes doesn't exist anymore :((( how can i transfer to new phone?

The iTunes feature now exists in macOS Finder.

1. Plug phone into computer with a usb/lighning cable
2. Phone will be listed in Finder under "Locations"

![Pasted_Image_12_16_22__16_12](https://user-images.githubusercontent.com/19795/208211540-ac781514-874d-4f11-8c8b-036c5ff3a24c.png)

**Comment by beaucollins** [comment]
If you use iCloud Keychain with two-factor enabled for your Apple account I find that they are now synced.

See https://github.com/mattrubin/Authenticator/issues/330#issuecomment-575311222

I have since moved to using macOS's built in two-factor system integrated into the system passwords.
